portfolio/src/main/java/com/google/sps/servlets/UserDataServlet.java

```diff
@@ -0,0 +1,204 @@
+        if (gameWon) {                 
+            long playerWins = (Long) pingPongEntity.getProperty("wins");
+            long winStreak = (Long) pingPongEntity.getProperty("winStreak");
+
+            playerWins++;
+            winStreak++;
+            
+            pingPongEntity.setProperty("wins", playerWins);
+            pingPongEntity.setProperty("winStreak", winStreak);
+            if (winStreak > (Long) pingPongEntity.getProperty("longestWinStreak")) {
+                pingPongEntity.setProperty("longestWinStreak", winStreak);
+            }
+        } else {
+            long playerLoses = (Long) pingPongEntity.getProperty("loses");
+            playerLoses++;
+            pingPongEntity.setProperty("loses", playerLoses);
+            pingPongEntity.setProperty("winStreak", 0);
+        }
+        datastore.put(pingPongEntity);
+    }
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        ptoString();
+        if (request.getParameter("query").equals("post")) {
+            doPost(request, response);
+            return;
+        } 
+
+        String getType = request.getParameter("getType");
+        if (getType.equals("login")) {
+            response.setContentType("application/json");
+
+            String thisUrl = "/games.html";
+            String responseJson;
+
+            if (!userService.isUserLoggedIn()) {
+                responseJson = createJsonForLog(false, thisUrl);
+                response.getWriter().println(responseJson);
+            } else if (userService.isUserLoggedIn()) {
+                confirmUserEntity();
+                responseJson = createJsonForLog(true, thisUrl);
+                response.getWriter().println(responseJson);
+            }
+        } else if (getType.equals("property")) {
+            response.setContentType("text/html");
+            if (request.getParameter("gameEntity").equals("PingPongEntity")) {
+                String propertyName = request.getParameter("propertyName");
+                String propertyVal = String.valueOf(pingPongEntity.getProperty(propertyName));
+                response.getWriter().println(propertyVal);
+            }
+        }
+    }
+
+    public String createJsonForLog(boolean loggedIn, String returnUrl) {
+        String returnLogUrl = (!loggedIn ? userService.createLoginURL(returnUrl) 
+        : userService.createLogoutURL(returnUrl));
+
+        String response = "{";
+        response += "\"loggedIn\": ";
+        response += "\"" + loggedIn + "\"";
+        response += ", ";
+        if (loggedIn) {
+            response += "\"nickname\": ";
+            Entity entity = getEntity("User");
+            response += "\"" + entity.getProperty("nickname")  + "\"";
+            response += ", ";
+        }
+        response += "\"returnUrl\": ";
+        response += "\"" + returnLogUrl + "\"";
+        response += "}";
+        return response;
+    }
```

Comment:
Maybe look into libraries that would build the json automatically for you, instead of building the json string yourself.  Doesn't gson have like a "toJson()" method?